### PR TITLE
Created a testing infrastructure for our scantron processing algorithm

### LINKED ISSUE #80: Comprehensive Testing for Scantron Processing Algorithm
We need to implement and ensure comprehensive testing of the Scantron processing algorithm. This algorithm handles critical tasks like extracting images from PDFs, aligning scantron images, detecting filled bubbles, and generating JSON outputs for the student results. To maintain reliability and accuracy, we need to cover both unit testing for individual components and integration testing to ensure everything works together smoothly.

**Requirements:**
Unit Testing: Write unit tests for each key function:
extractImagesFromPdf: Test if images are correctly extracted from PDFs, saved with the right dimensions, and named properly.
align_image: Check alignment accuracy and behavior when images are distorted or missing key points.
crop_roi: Validate that the ROIs are properly cropped and extracted for different sections (e.g., student ID, answers).
get_responses_bubble_row and bubble_column: Ensure the bubble detection is accurate for various fill levels (light, overfilled, smudged) and edge cases (no bubbles filled).
student_id: Test if student IDs are correctly extracted from the bubbles and test behavior for incomplete or incorrect entries.

**Integration Testing:**
Test the full workflow, starting from PDF uploads to the final JSON generation.
Verify that multiple scantron sheets can be processed together without errors.
Ensure the system handles edge cases like missing bubbles, skewed images, or poor image quality.

**Performance Testing:**
Might be interesting to see if the program has the same, or any, success with the 200 dpi as the 600 dpi. Since the printer at slu has different quality options for scanning.
Measure processing time for large PDF files and ensure that the system can handle multiple pages efficiently.
Stress test with various input sizes and see how the system performs under load.

**Error Handling:**
Test how the system behaves with corrupted PDFs, incomplete images.
Ensure that the system doesn't crash when encountering problematic files.

You can document the results too for sharing the performance with everyone.

### PR DESCRIPTION
**Pull request by ramezmosad**
Fixes #80 

**What was changed?**

I have added comprehensive unit and integration tests for the Scantron processing algorithm, specifically targeting the Scantron95945 class. These tests are written using the Pytest framework and cover key functions such as extractImagesFromPdf, align_image, crop_roi, get_responses_bubble_row, bubble_column, student_id, and the overall integration of these components. The new test scripts aim to validate the correctness, reliability, and robustness of the algorithm under various conditions.

**Why was it changed?**

The change was made to address the need for thorough testing of the Scantron processing algorithm to ensure its reliability and accuracy. The algorithm performs critical tasks like extracting images from PDFs, aligning scantron images, detecting filled bubbles, and generating JSON outputs with student results. By implementing these tests, we aim to catch potential bugs, handle edge cases (such as corrupted PDFs, distorted images, and varying fill levels of bubbles), and ensure that the system can handle real-world scenarios without errors.

**How was it changed?**

I added several test classes and functions within the tests directory to cover both unit and integration testing:

*Unit Tests*:
1. *test_align.py*: 
* Tests the align_image function for correct image alignment and handling cases where keypoints are missing.
* Verifies that images are correctly aligned when features are present.
* Checks that the function returns the original image when keypoints are absent.

2. *test_bubble_detection.py*: 
* Tests get_responses_bubble_row and bubble_column functions for accurate bubble detection under various fill levels. 
* Ensures single and multiple filled bubbles are detected correctly. 
* Validates behavior when no bubbles are filled.

3. *test_crop_ROI.py*: 
* Tests the crop_roi function to ensure ROIs are correctly cropped and extracted. 
* Validates proper cropping when all markers are present. 
* Uses synthetic images with markers to simulate different scenarios.

4. *test_id.py*: 
* Tests the student_id function for accurate extraction of student IDs.
* Checks for complete IDs with all digits filled. 
* Tests behavior with incomplete IDs, ensuring missing digits are represented as 'X'.

5. *test_extract.py*: 
* Tests extractImagesFromPdf for correct image extraction from PDFs.
* Mocks PDF files to simulate extraction without relying on actual files. 
* Validates that images are saved with correct dimensions and naming conventions. 

6. *test_error_handling.py*: 
* Tests how the system handles errors such as corrupted PDFs or missing image files.
* Ensures the system doesn't crash and logs appropriate error messages.

*Integration Test*:
1. *test_workflow.py*: 
* Simulates the full workflow from PDF upload to JSON generationMocks methods interacting with external systems or files. 
* Verifies that the system correctly processes PDFs and outputs the expected JSON structure.

*Helper Class*:
Scantron95945TestHelper: A subclass of Scantron95945 used to adjust certain methods for testing purposes, such as lowering thresholds in get_responses_bubble_row.

**Other Comments**:
I've also manually tested our program with scantron sheets with two different image qualities: one at 200dpi, and the other at 600dpi. The accuracy for both image qualities was perfect (100%).

All of the unit tests I wrote pass:
Screenshot 2024-09-29 at 11 50 11 PM

### CODE REVIEW

File: BubbleScan-AI/tests/test_crop_ROI.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+class TestCropROI:
+    @patch('cv2.imwrite', return_value=True)
+    @patch('os.makedirs')
+    @patch('cv2.imread')
+    @patch.object(Scantron95945, 'align_image')
+    @patch.object(Scantron95945, 'extractROIs')
+    @patch.object(Scantron95945, 'template_matching')
+    @patch.object(Scantron95945, 'extractImagesFromPdf')
+    def test_valid_crop(self, mock_extractImagesFromPdf, mock_template_matching, mock_extractROIs, mock_align_image, mock_imread, mock_makedirs, mock_imwrite):
```
Comment (mylescorrea) [inline BubbleScan-AI/tests/test_crop_ROI.py:19]: potentially long parameter list?

**Review by mylescorrea — COMMENTED** [review]
Good job with the comments and readability with the tests! Everything looks nicely organized.

**Comment by kate-holdener** [comment]
@ramezmosad this is an impressive test suite! When I run the tests, they are not finding the PDF directory. Can you share how you run these tests (from which directory and what environment variables do you set)?

File: BubbleScan-AI/data/ROIs/Image_3/student_id_roi.jpg
Comment (Chirag2x) [inline BubbleScan-AI/data/ROIs/Image_3/student_id_roi.jpg:1]: It's really impressive, You took the efforts to generate the different sections for Testing!

File: BubbleScan-AI/data/ROIs/Image_7/second_column_roi.jpg
Comment (Chirag2x) [inline BubbleScan-AI/data/ROIs/Image_7/second_column_roi.jpg:1]: Good Even considered the empty cases

File: BubbleScan-AI/tests/test_bubble_detection.py
Comment (Chirag2x) [inline BubbleScan-AI/tests/test_bubble_detection.py:1]: This is in detail test, The comments are super helpful

File: BubbleScan-AI/tests/test_workflow.py
Comment (Chirag2x) [inline BubbleScan-AI/tests/test_workflow.py:1]: This is a good!

**Review by Chirag2x — COMMENTED** [review]
It is well Structured work, the comments are really helpful, All the test cases are running properly has intended. You have taken error handling and all the other outliers cases in to consideration which was necessary to get the good performance. This is an amazing work @ramezmosad. Keep It up.!!
Also try to change this .gitignore file since pylint is failing.
Screenshot 2024-10-02 at 12 02 44 PM

**Review by Chirag2x — APPROVED** [review]
